File: API/ApiModels/LeakModel.cs
using System.ComponentModel.DataAnnotations;

namespace CredentialLeakageMonitoring.API.ApiModels
{
    public record LeakModel
    {
        [Required]
        public Guid Id { get; init; }

        [Required]
        public string EmailHash { get; init; } = string.Empty;

        [Required]
        public string ObfuscatedPassword { get; init; } = string.Empty;

        [Required]
        public DateTimeOffset FirstSeen { get; init; }

        [Required]
        public DateTimeOffset LastSeen { get; init; }

        [Required]
        public string Domain { get; init; } = string.Empty;

        [Required]
        public List<CustomerModel> AssociatedCustomers { get; init; } = [];
    }
}

File: WebApp/lib/models/leak_model.dart
import 'package:json_annotation/json_annotation.dart';
import 'customer_model.dart';

part 'leak_model.g.dart';

@JsonSerializable()
class LeakModel {
  LeakModel({
    required this.id,
    required this.emailHash,
    required this.firstSeen,
    required this.lastSeen,
    required this.domain,
    required this.associatedCustomers,
    required this.passwordHash,
  });

  factory LeakModel.fromJson(Map<String, dynamic> json) =>
      _$LeakModelFromJson(json);

  final List<CustomerModel> associatedCustomers;
  final String domain;
  final String emailHash;
  final DateTime firstSeen;
  final String id;
  final DateTime lastSeen;
  final String passwordHash;

  Map<String, dynamic> toJson() => _$LeakModelToJson(this);
}
